### PR TITLE
Add support to decode a resume token.

### PR DESCRIPTION
**Pull request by tony-zfs**
Adding a new option -t resume_token within zstream dump. This
now allows users to decode a resume token to retrieve the toname
field. This can be useful for tools that need this information.

Signed-off-by: Tony Perkins <tperkins@datto.com>

This provides a way to extract more information from a resume token.

### Motivation and Context
This provides a way for us to generically identify a failed zfs send/recv by 
inspecting the resume token.

### Description
This uses libzfs to parse the resume token, and display the contents from 
the nvlist.

### How Has This Been Tested?
A new functional test has been created under rsend to help test the new option.
This was tested under Ubuntu / ZoL, and our system build test infrastructure.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X ] Documentation (a change to man pages or other documentation)

### Checklist:
- [ X] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ X] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [X ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [X ] I have run the ZFS Test Suite with this change applied.
- [X ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by DeHackEd** [comment]
FYI you can also do `zfs send -t $TOKEN -v -n` to see the same information. If the current host isn't the correct one to do the send it will also produce an error trying to execute on the token, but it will still produce the output.

This zstreamdump idea is cleaner though.

**Comment by ahrens** [comment]
I think it would be cleaner for this to be a new subcommand of `zstream`, since it `zstream dump` takes a stream as input.  Perhaps `zstream token <TOKEN>`.

**Comment by tony-zfs** [comment]
Updated PR to use new subcommand "token" per Matt's advice.

File: cmd/zstream/zstream_token.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	if (argc < 2) {
+		(void) fprintf(stderr, "Need to pass the resume token\n");
+		zstream_usage();
+		return (-1);
+	}
+
+	resume_token = argv[1];
+
+	if (resume_token) {
```
Comment (ahrens) [inline cmd/zstream/zstream_token.c:61]: I don't think it can be NULL, since you already checked `argc` above.

File: cmd/zstream/zstream_token.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		libzfs_handle_t *hdl = libzfs_init();
+
+		nvlist_t *resume_nvl =
+		    zfs_send_resume_token_to_nvlist(hdl, resume_token);
+		libzfs_fini(hdl);
+
+		if (resume_nvl == NULL) {
+			/*
+			 * zfs_error_aux has already been set by
+			 * zfs_send_resume_token_to_nvlist()
```
Comment (ahrens) [inline cmd/zstream/zstream_token.c:71]: I don't see where the zfs_error_aux gets printed?

File: man/man8/zstream.8 [hunk 3]
```diff
@@ -67,6 +70,12 @@ Implies verbose.
 .El
 .It Xo
 .Nm
+.Cm token
+.Ar resume_token 
+.Xc
+Dumps  zfs resume token information
```
Comment (ahrens) [inline man/man8/zstream.8:76]:
```suggestion
Dumps zfs resume token information 
```

File: man/man8/zstreamdump.8 [hunk 4]
```diff
@@ -32,6 +33,12 @@ The following options are supported:
 Suppress the validation of checksums.
 .RE
 
+.ad
+.sp .6
+.RS 4n
+Only dump information about the resume token.
+.RE
```
Comment (ahrens) [inline man/man8/zstreamdump.8:40]: I think you can revert the changes to this file since it's now (only) in `zstream.8`, right?

File: cmd/zstream/zstream_dump.c [hunk 5]
```diff
@@ -38,6 +39,9 @@
 #include <unistd.h>
 #include <stddef.h>
 
+#include <libzfs.h>
+#include <libzfs_core.h>
```
Comment (ahrens) [inline cmd/zstream/zstream_dump.c:43]: Is this "code cleanup" (since something in this header uses stuff from libzfs[_core]), or an inadvertent change?

File: man/man8/zstreamdump.8 (hunk 4 above)
Comment (tony-zfs) [reply man/man8/zstreamdump.8:40]: Yep.

File: cmd/zstream/zstream_dump.c (hunk 5 above)
Comment (tony-zfs) [reply cmd/zstream/zstream_dump.c:43]: Left over, I'll get rid of it.

**Comment by tony-zfs** [comment]
Cleaned up the old changes.

File: cmd/zstream/zstream_token.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+	libzfs_handle_t *hdl = libzfs_init();
+
+	nvlist_t *resume_nvl =
+	    zfs_send_resume_token_to_nvlist(hdl, resume_token);
+	libzfs_fini(hdl);
+
+	if (resume_nvl == NULL) {
+		(void) fprintf(stderr,
+		    "Unable to parse resume token\n");
```
Comment (ahrens) [inline cmd/zstream/zstream_token.c:69]: I think it would be helpful to print the aux error (which can be obtained via `libzfs_error_description()`), since zfs_send_resume_token_to_nvlist() does provide more descriptive messages in it.

Comment (tony-zfs) [reply cmd/zstream/zstream_token.c]: updated.

File: tests/zfs-tests/tests/functional/rsend/send-c_zstreamdump.ksh [hunk 7]
```diff
@@ -26,6 +27,9 @@
 # 1. Create a full compressed send stream
 # 2. Verify zstreamdump shows this stream has the relevant features
 # 3. Verify zstreamdump's accounting of logical and compressed size is correct
+# 4. Verify the toname from a resume token
+# 5. Verify it fails with corrupted resume token
+# 6. Verify it fails wiith missing resume token
```
Comment (PaulZ-98) [inline tests/zfs-tests/tests/functional/rsend/send-c_zstreamdump.ksh:32]: typo wiith -> with

File: tests/zfs-tests/tests/functional/rsend/rsend.kshlib [hunk 8]
```diff
@@ -657,6 +658,21 @@ function resume_test
...
+{
+	sendcmd=$1
+	streamfs=$2
+	recvfs=$3
+
+	log_must eval "$sendcmd > /$streamfs/1"
+	mess_send_file /$streamfs/1
+	log_mustnot zfs recv -suv $recvfs < /$streamfs/1 2>&1
+	token=$(zfs get -Hp -o value receive_resume_token $recvfs)
+        echo "$token" > /$streamfs/resume_token
```
Comment (PaulZ-98) [inline tests/zfs-tests/tests/functional/rsend/rsend.kshlib:671]: fix indent on echo line

File: cmd/zstream/zstream_token.c [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+int
+zstream_do_token(int argc, char *argv[])
+{
+	char *resume_token = NULL;
+
+	if (argc < 2) {
+		(void) fprintf(stderr, "Need to pass the resume token\n");
+		zstream_usage();
+		return (-1);
```
Comment (behlendorf) [inline cmd/zstream/zstream_token.c:56]: The `return (-1)` here is unreachable.  Let's remove it to be consistent with how it's used elsewhere in `zstream`.

File: cmd/zstream/zstream_token.c [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+	nvlist_t *resume_nvl =
+	    zfs_send_resume_token_to_nvlist(hdl, resume_token);
+
+	if (resume_nvl == NULL) {
+		(void) fprintf(stderr,
+		    "Unable to parse resume token: %s\n",
+		    libzfs_error_description(hdl));
+		libzfs_fini(hdl);
+		return (-2);
```
Comment (behlendorf) [inline cmd/zstream/zstream_token.c:71]: This should be `return (1)` to avoid a negative exit code.  `zstream` should return 0 on success and 1 on error, like the `zfs` and `zpool` commands.

Comment (tony-zfs) [reply cmd/zstream/zstream_token.c]: fixed

File: tests/zfs-tests/tests/functional/rsend/rsend.kshlib (hunk 8 above)
Comment (tony-zfs) [reply tests/zfs-tests/tests/functional/rsend/rsend.kshlib]: fixed

File: tests/zfs-tests/tests/functional/rsend/send-c_zstreamdump.ksh (hunk 7 above)
Comment (tony-zfs) [reply tests/zfs-tests/tests/functional/rsend/send-c_zstreamdump.ksh]: fixed

File: cmd/zstream/zstream_token.c (hunk 9 above)
Comment (tony-zfs) [reply cmd/zstream/zstream_token.c]: got it.
